invisible_cities/cities/psf_liquidata.py

```diff
@@ -0,0 +1,193 @@
+@city
+def create_psf_city( files_in, file_out, compression, event_range
+                   , detector_db, run_number
+                   , xrange      : Sequence[float]
+                   , yrange      : Sequence[float]
+                   , zbins       : Sequence[float]
+                   , xsectors    : Sequence[float]
+                   , ysectors    : Sequence[float]
+                   , bin_size_xy : float           = 1 * units.mm
+                   ):
+    sipms   = DataSiPM(detector_db, run_number)
+    ranges  = xrange  , yrange
+    sectors = xsectors, ysectors
+
+    nbin      = (np.diff(ranges, axis=1) / bin_size_xy).astype(int).flatten()
+    bin_edges = [np.linspace(*limits, n+1) for limits, n in zip(ranges, nbin)]
+
+    columns_to_drop     = "Xrms Yrms Qc Ec".split()
+    drop_unused_columns = get.dst    *      partial(column_dropper, columns_to_drop)   >> put.dst
+    create_slices_z     = get.dst    * flat(partial(split_in_z, zbins, ranges, sipms)) >> put.z.dst_z
+    create_sectors_xy   = get.dst_z  * flat(partial(split_in_xy, sectors, bin_edges))  >> put.x.y.dst_xy
+    compute_psf         = get.dst_xy *      partial(psf_builder, bin_edges)            >> put.factors.entries.centers
+    build_dataframe     = get.x.y.z.factors.entries.centers * df_builder               >> put.psf
+
+    pipeline = pipe( drop_unused_columns
+                   , create_slices_z
+                   , create_sectors_xy
+                   , compute_psf
+                   , build_dataframe
+                   , get.psf
+                   , out(combine_psfs, None)
+                   )
+
+    result = pipeline(dst_from_files(files_in, "RECO", "Events"))
+
+    with tb.open_file(file_out, 'w') as outfile:
+        write_psf = psf_writer(outfile, compression=compression)
+
+        # TODO: Probably better to ignore this writer
+        # and use df_writer directly
+        for (x, y, z), df in result.groupby(["x", "y", "z"]):
+            xr      = df.xr.drop_duplicates()
+            yr      = df.yr.drop_duplicates()
+            zr      = df.zr.drop_duplicates()
+            shape   = len(xr), len(yr), len(zr)
+            factor  = df.factor .values.reshape(shape)
+            entries = df.entries.values.reshape(shape)
+
+            write_psf( xr, yr, zr
+                     , x , y , z
+                     , factor
+                     , entries
+                     )
+
+    return result
+
+
+from typing import List
+from typing import Dict
+from typing import Iterator
+
+from argparse  import Namespace
+
+from ..io.dst_io import load_dst
+
+def dst_from_files(paths: List[str], group: str, node:str) -> Iterator[Dict[str, pd.DataFrame]]:
+    """
+    Reader for a generic dst.
+    """
+    for path in paths:
+        try:
+            df = load_dst(path, group, node)
+        except tb.exceptions.NoSuchNodeError:
+            continue
+
+        yield Namespace(dst = df)
+
+def column_dropper(columns, df):
```

Comment:
This (and the next 4-ish functions) seem to be written for use just here, and they are all used with `partial`. You should probably curry them.

Reply:
I kinda like partial, but ok.

What are your thoughts on adding a `curryable` decorator? Something like
```
from inspect import signature

def curryable(f):
    total_args = len(signature(f).parameters)
    def g(*args):
        if len(args) < total_args:
            return curryable(partial(f, *args))
        else:
            return f(*args)
    returng g
```

I am afraid it might be a bit obscure to find a function _in python_ that is called with too few parameters, but it would avoid writing these functions that essentially implement a partial application.

Reply:
I dislike `partial` because

+ It wastes a lot of precious space
+ It puts the callable along with its arguments inside an argument list. This sends a very misleading message, and it takes careful conscious effort to suppress one's natural parsing instincts.

`curryable` looks OKish.

+ `@wraps(f)` is missing
+ I'm always wary of `inspect`, because it's easy to make assumptions which don't work out in the general case. This version completely ignores the possibility of keyword arguments. We'd need to think to what extent it is worth the effort to cover all cases.
+ `partially_applicable` would be a more correctly-descriptive name.
+ Stick it in an inner loop, and it might degrade performance noticeably.

> I am afraid it might be a bit obscure to find a function in python that is called with too few parameters

?

[Aside, point of pedantic nomenclature: '... called with too few *arguments*'. Parameters are *names*; arguments are *values*. I know that people tend to use them interchangeably, and in 70% of cases it doesn't matter (because parameters are bound to arguments and therefore represent or 'contain' them). But in the 30% of the cases where it does matter, it really helps if everyone is already in the habit of using the correct terminology.]

Anyway ... I guess you're contrasting it with 

+ `partial` which is in-your-face at the point of use
+ curried-by-hand functions (e.g. `operator.itemgetter`) which strictly separate early and late arguments

so you're worried that `curryable` (or `partially_applicable`) will confuse people because, at point of use,

+ it's excessively flexible in what it accepts
+ whether it runs or not is implicitly determined by how many arguments you pass.

Reply:
> I dislike partial because [...]

I agree on both points.

> I'm always wary of inspect, because it's easy to make assumptions which don't work out in the general case. This version completely ignores the possibility of keyword arguments. We'd need to think to what extent it is worth the effort to cover all cases.

Indeed. This was just a draft to get your view on the mere idea of adding this kind of function. But also...

> Stick it in an inner loop, and it might degrade performance noticeably.

I was thinking of using this strictly for city components, so we get rid of the pattern

```
def outer(*init_args):
    ...
    def inner(*late_args):
        ...
    return inner
```

which, IMO, becomes a bit noisy. Also notice that it might not be always optimal. For instance, if `init_args` need to be massaged into something more useful, it doesn't make sense to do it for every call of the inner function.

> it really helps if everyone is already in the habit of using the correct terminology.

yeah.. I'm sorry, I've always had trouble using the correct terminology... as you probably noticed :) I'll keep trying

> ?
> so you're worried that curryable (or partially_applicable) will confuse people because, at point of use,
>
>    it's excessively flexible in what it accepts
>    whether it runs or not is implicitly determined by how many arguments you pass.

Yes, mostly that. Also, it is very easy for people (particularly those less interested in code development) to ignore the decorator and think that they don't understand the code.

Reply:
> I was thinking of using this strictly for city components, so we get rid of the pattern
> 
> ``` python
> def outer(*init_args):
>     ...
>     def inner(*late_args):
>         ...
>     return inner
> ```
> 
> which, IMO, becomes a bit noisy.

Agreed: this pattern is a noisy PITA, and it would be good to abstract it. If only we had Lisp macros. Which we do in Julia.

> it is very easy for people (particularly those less interested in code development) to ignore the decorator and think that they don't understand the code.

And you think that the nested-def pattern is better in this respect?

Reply:
In my experience, nested functions seem to create some confusion as well. In practice, if someone wants to write a city, they will probably get some IC 101 course anyway so perhaps it doesn't matter. I think a simple demonstration of how `partially_applicable` works is as good as explaining nested-defs.